File: internal/api/details.go
package api

import (
	"encoding/json"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/rmrobinson/cupola/internal/domain"
)

func (h *Handler) getDetail(w http.ResponseWriter, r *http.Request) {
	dt := domain.DomainType(chi.URLParam(r, "domain"))
	id := r.URL.Query().Get("id")
	if id == "" {
		http.NotFound(w, r)
		return
	}

	detail, ok := h.detailFor(dt, id)
	if !ok {
		http.NotFound(w, r)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(detail)
}

func (h *Handler) detailFor(dt domain.DomainType, id string) (domain.Detail, bool) {
	state := h.store.Get(dt)
	if state == nil {
		return domain.Detail{}, false
	}

	switch dt {
	case domain.DomainWeatherAlerts:
		s, ok := state.(domain.WeatherAlerts)
		if !ok {
			return domain.Detail{}, false
		}
		for _, a := range s.Alerts {
			if a.ID != id {
				continue
			}
			return domain.Detail{
				Domain:      string(dt),
				ID:          a.ID,
				Title:       a.Title,
				Severity:    string(a.Severity),
				Description: a.Summary,
				Fields: []domain.DetailField{
					fieldTime("onset", &a.Onset),
					fieldTime("expires", &a.Expires),
				},
				SourceURL: safeSourceURL(a.SourceURL),
			}, true
		}

	case domain.DomainTransitAlerts:
		s, ok := state.(domain.TransitAlerts)
		if !ok {
			return domain.Detail{}, false
		}
		for _, a := range s.Alerts {
			if a.ID != id {
				continue
			}
			fields := []domain.DetailField{
				{Key: "agency_id", Value: a.AgencyID},
				fieldTime("starts_at", a.StartsAt),
				fieldTime("ends_at", a.EndsAt),
			}
			if len(a.AffectedRoutes) > 0 {
				fields = append(fields, domain.DetailField{Key: "affected_routes", Value: strings.Join(a.AffectedRoutes, ", ")})
			}
			return domain.Detail{
				Domain:      string(dt),
				ID:          a.ID,
				Title:       a.Title,
				Subtitle:    a.AgencyID,
				Severity:    string(a.Severity),
				Description: a.Description,
				Fields:      compactFields(fields),
			}, true
		}

	case domain.DomainMunicipalAlerts:
		s, ok := state.(domain.MunicipalAlerts)
		if !ok {
			return domain.Detail{}, false
		}
		for _, a := range s.Alerts {
			if a.ID != id {
				continue
			}
			fields := []domain.DetailField{
				{Key: "source_id", Value: a.SourceID},
				{Key: "alert_type", Value: a.AlertType},
				fieldStringPtr("area", a.Area),
				fieldTime("starts_at", a.StartsAt),
				fieldTime("ends_at", a.EndsAt),
				fieldTime("published_at", &a.PublishedAt),
			}
			sourceURL := ""
			if a.URL != nil {
				sourceURL = safeSourceURL(*a.URL)
			}
			return domain.Detail{
				Domain:      string(dt),
				ID:          a.ID,
				Title:       a.Title,
				Subtitle:    valueOrEmpty(a.Area),
				Severity:    string(a.Severity),
				Description: a.Description,
				Fields:      compactFields(fields),
				SourceURL:   sourceURL,
			}, true
		}

	case domain.DomainFlagStatus:
		s, ok := state.(domain.FlagStatus)
		if !ok || id != "current" {
			return domain.Detail{}, false
		}
		return domain.Detail{
			Domain:      string(dt),
			ID:          id,
			Title:       "flag.status",
			Description: valueOrEmpty(s.Reason),
			Fields: compactFields([]domain.DetailField{
				fieldBool("at_half_mast", s.AtHalfMast),
				fieldTime("since", s.Since),
				fieldTime("until", s.Until),
			}),
			SourceURL: safeSourceURL(s.SourceURL),
		}, true

	case domain.DomainTrafficIncidents:
		s, ok := state.(domain.TrafficIncidents)
		if !ok {
			return domain.Detail{}, false
		}
		for _, inc := range s.Incidents {
			if inc.ID != id {
				continue
			}
			var loc *domain.DetailLocation
			if !inc.ApproximateLocation && (inc.Lat != 0 || inc.Lon != 0) {
				loc = &domain.DetailLocation{Lat: inc.Lat, Lon: inc.Lon}
			}
			locationValue := "specific"
			if inc.ApproximateLocation {
				locationValue = inc.LocationLabel
				if locationValue != "" {
					locationValue += " (approximate; not shown on map)"
				} else {
					locationValue = "approximate; not shown on map"
				}
			}
			return domain.Detail{
				Domain:      string(dt),
				ID:          inc.ID,
				Title:       inc.Type,
				Subtitle:    inc.RoadName,
				Severity:    inc.Severity,
				Description: inc.Description,
				Fields: compactFields([]domain.DetailField{
					{Key: "road_name", Value: inc.RoadName},
					{Key: "type", Value: inc.Type},
					{Key: "location", Value: locationValue},
					fieldTime("starts_at", inc.StartsAt),
					fieldTime("ends_at", inc.EndsAt),
				}),
				SourceURL: safeSourceURL(inc.SourceURL),
				Location:  loc,
			}, true
		}

	case domain.DomainWaterwayConditions:
		s, ok := state.(domain.WaterwayConditions)
		if !ok {
			return domain.Detail{}, false
		}
		for _, g := range s.Gauges {
			if g.ID != id {
				continue
			}
			return domain.Detail{
				Domain:      string(dt),
				ID:          g.ID,
				Title:       g.Name,
				Subtitle:    g.WaterwayName,
				Severity:    g.AdvisoryStatus,
				Description: valueOrEmpty(g.AdvisoryText),
				Fields: compactFields([]domain.DetailField{
					{Key: "waterway_name", Value: g.WaterwayName},
					fieldFloatPtr("level_m", g.LevelM, "m"),
					fieldFloatPtr("flow_cms", g.FlowCMS, "m3_per_s"),
					fieldFloatPtr("temp_c", g.TempC, "celsius"),
					{Key: "advisory_status", Value: g.AdvisoryStatus},
					fieldTime("updated_at", &g.UpdatedAt),
				}),
				SourceURL: safeSourceURL(g.SourceURL),
				Location:  &domain.DetailLocation{Lat: g.Lat, Lon: g.Lon},
			}, true
		}
	}

	return domain.Detail{}, false
}

func fieldStringPtr(key string, value *string) domain.DetailField {
	if value == nil {
		return domain.DetailField{}
	}
	return domain.DetailField{Key: key, Value: *value}
}

func fieldTime(key string, value *time.Time) domain.DetailField {
	if value == nil || value.IsZero() {
		return domain.DetailField{}
	}
	return domain.DetailField{Key: key, Value: value.Format(time.RFC3339)}
}

func fieldBool(key string, value bool) domain.DetailField {
	return domain.DetailField{Key: key, Value: value}
}

func fieldFloatPtr(key string, value *float64, unit string) domain.DetailField {
	if value == nil {
		return domain.DetailField{}
	}
	return domain.DetailField{Key: key, Value: *value, Unit: unit}
}

func compactFields(fields []domain.DetailField) []domain.DetailField {
	out := make([]domain.DetailField, 0, len(fields))
	for _, f := range fields {
		if f.Key != "" && f.Value != nil && f.Value != "" {
			out = append(out, f)
		}
	}
	return out
}

func valueOrEmpty(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}

func safeSourceURL(raw string) string {
	if raw == "" {
		return ""
	}
	u, err := url.Parse(raw)
	if err != nil || u.Host == "" {
		return ""
	}
	if u.Scheme != "http" && u.Scheme != "https" {
		return ""
	}
	return raw
}

File: cmd/cupola/frontend/js/details.js
(function () {
  'use strict';

  let root = null;
  let lastFocus = null;
  let activeController = null;
  let requestSeq = 0;

  const FIELD_LABELS = {
    advisory_status: 'Advisory',
    affected_routes: 'Affected routes',
    agency_id: 'Agency',
    alert_type: 'Type',
    area: 'Area',
    at_half_mast: 'At half-mast',
    ends_at: 'Ends',
    expires: 'Expires',
    flow_cms: 'Flow',
    level_m: 'Level',
    location: 'Location',
    onset: 'Onset',
    published_at: 'Published',
    road_name: 'Road',
    since: 'Since',
    source_id: 'Source',
    starts_at: 'Starts',
    temp_c: 'Temperature',
    type: 'Type',
    until: 'Until',
    updated_at: 'Updated',
    waterway_name: 'Waterway',
  };

  const UNIT_SYMBOLS = {
    celsius: '°C',
    m: 'm',
    m3_per_s: 'm³/s',
  };

  const DOMAIN_LABELS = {
    'flag.status': 'Flag status',
    'municipal.alerts': 'Municipal alert',
    'traffic.incidents': 'Traffic incident',
    'transit.alerts': 'Transit alert',
    'waterway.conditions': 'Waterway',
    'weather.alerts': 'Weather alert',
  };

  function esc(s) {
    return String(s || '').replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
  }

  function escAttr(s) {
    return esc(s).replace(/"/g, '&quot;');
  }

  function fmtValue(value) {
    if (typeof value === 'boolean') return value ? 'Yes' : 'No';
    if (!value) return '';
    const ts = Date.parse(value);
    if (!Number.isNaN(ts) && /^\d{4}-\d{2}-\d{2}T/.test(value)) {
      return new Date(ts).toLocaleString(undefined, {
        month: 'short', day: 'numeric', hour: 'numeric', minute: '2-digit',
      });
    }
    return value;
  }

  function labelFor(key) {
    return FIELD_LABELS[key] || key.replace(/_/g, ' ');
  }

  function unitFor(unit) {
    return UNIT_SYMBOLS[unit] || unit || '';
  }

  function titleCase(s) {
    return String(s || '')
      .replace(/[_-]/g, ' ')
      .replace(/\w\S*/g, word => word.charAt(0).toUpperCase() + word.slice(1));
  }

  function fieldValue(d, key) {
    const f = (d.fields || []).find(field => field.key === key);
    return f ? f.value : null;
  }

  function titleFor(d) {
    if (d.domain === 'flag.status') {
      return fieldValue(d, 'at_half_mast') ? 'Flag at half-mast' : 'Flag at full mast';
    }
    if (d.domain === 'traffic.incidents') return titleCase(d.title);
    return d.title || DOMAIN_LABELS[d.domain] || 'Details';
  }

  function safeHTTPURL(raw) {
    if (!raw) return '';
    try {
      const u = new URL(raw, window.location.href);
      if (u.protocol !== 'http:' && u.protocol !== 'https:') return '';
      return u.href;
    } catch {
      return '';
    }
  }

  function isOpen() {
    return root && !root.classList.contains('hidden');
  }

  function focusableElements() {
    if (!root) return [];
    return [...root.querySelectorAll('a[href], button:not([disabled]), [tabindex]:not([tabindex="-1"])')]
      .filter(el => el.offsetParent !== null);
  }

  function ensureRoot() {
    if (root) return root;
    root = document.createElement('div');
    root.id = 'detail-root';
    root.className = 'detail-root hidden';
    root.setAttribute('aria-hidden', 'true');
    document.body.appendChild(root);

    root.addEventListener('click', e => {
      if (e.target.classList.contains('detail-backdrop') ||
          e.target.closest('.detail-close')) {
        close();
      }
    });

    document.addEventListener('keydown', e => {
      if (!isOpen()) return;
      if (e.key === 'Escape') {
        close();
        return;
      }
      if (e.key !== 'Tab') return;
      const focusable = focusableElements();
      if (focusable.length === 0) {
        e.preventDefault();
        return;
      }
      const first = focusable[0];
      const last = focusable[focusable.length - 1];
      if (e.shiftKey && document.activeElement === first) {
        e.preventDefault();
        last.focus();
      } else if (!e.shiftKey && document.activeElement === last) {
        e.preventDefault();
        first.focus();
      }
    });

    return root;
  }

  function triggerFromEvent(e) {
    const explicit = e.target.closest('.detail-open-action');
    const trigger = explicit || e.target.closest('[data-detail-domain][data-detail-id]');
    if (!trigger) return;
    if (!explicit && e.target.closest('a,button,input,select,textarea')) return;
    open(trigger.dataset.detailDomain, trigger.dataset.detailId);
  }

  function renderShell(content) {
    const el = ensureRoot();
    el.innerHTML = `
      <div class="detail-backdrop"></div>
      <aside class="detail-drawer" role="dialog" aria-modal="true" aria-label="Details">
        ${content}
      </aside>
    `;
    el.classList.remove('hidden');
    el.setAttribute('aria-hidden', 'false');
    el.querySelector('.detail-close')?.focus();
  }

  function renderLoading() {
    renderShell(`
      <div class="detail-header">
        <div>
          <div class="detail-kicker">Details</div>
          <h2>Loading</h2>
        </div>
        <button class="detail-close" type="button" aria-label="Close">&times;</button>
      </div>
      <div class="detail-body"><p class="detail-muted">Loading details...</p></div>
    `);
  }

  function renderError(message) {
    renderShell(`
      <div class="detail-header">
        <div>
          <div class="detail-kicker">Details</div>
          <h2>Unavailable</h2>
        </div>
        <button class="detail-close" type="button" aria-label="Close">&times;</button>
      </div>
      <div class="detail-body"><p class="detail-error">${esc(message)}</p></div>
    `);
  }

  function renderDetail(d) {
    const severity = d.severity && d.severity !== 'none'
      ? `<span class="detail-severity">${esc(d.severity)}</span>` : '';
    const fields = (d.fields || []).map(f => {
      const unit = unitFor(f.unit);
      const value = `${fmtValue(f.value)}${unit ? ` ${unit}` : ''}`;
      return `
      <div class="detail-field">
        <dt>${esc(labelFor(f.key))}</dt>
        <dd>${esc(value)}</dd>
      </div>
    `;
    }).join('');
    const location = d.location
      ? `<div class="detail-field"><dt>Location</dt><dd>${Number(d.location.lat).toFixed(5)}, ${Number(d.location.lon).toFixed(5)}</dd></div>`
      : '';
    const sourceURL = safeHTTPURL(d.source_url);
    const source = sourceURL
      ? `<a class="detail-source" href="${escAttr(sourceURL)}" target="_blank" rel="noopener">Open source page</a>`
      : '';

    renderShell(`
      <div class="detail-header">
        <div>
          <div class="detail-kicker">${esc(DOMAIN_LABELS[d.domain] || d.domain || 'Details')}</div>
          <h2>${esc(titleFor(d))}</h2>
          ${d.subtitle ? `<div class="detail-subtitle">${esc(d.subtitle)}</div>` : ''}
        </div>
        <button class="detail-close" type="button" aria-label="Close">&times;</button>
      </div>
      <div class="detail-body">
        ${severity}
        ${d.description ? `<p class="detail-description">${esc(d.description)}</p>` : ''}
        ${fields || location ? `<dl class="detail-fields">${fields}${location}</dl>` : ''}
        ${source}
      </div>
    `);
  }

  async function open(domain, id) {
    if (!domain || !id) return;
    const seq = ++requestSeq;
    if (activeController) activeController.abort();
    activeController = new AbortController();
    lastFocus = document.activeElement;
    renderLoading();
    try {
      const url = `/api/v1/details/${encodeURIComponent(domain)}?id=${encodeURIComponent(id)}`;
      const res = await fetch(url, { signal: activeController.signal });
      if (!res.ok) throw new Error(`Details request failed (${res.status})`);
      const detail = await res.json();
      if (seq !== requestSeq) return;
      renderDetail(detail);
    } catch (err) {
      if (err.name === 'AbortError') return;
      if (seq !== requestSeq) return;
      console.error('[cupola] detail fetch failed', err);
      renderError('Details are not available for this item.');
    }
  }

  function close() {
    if (!root) return;
    requestSeq++;
    if (activeController) activeController.abort();
    activeController = null;
    root.classList.add('hidden');
    root.setAttribute('aria-hidden', 'true');
    root.innerHTML = '';
    if (lastFocus && typeof lastFocus.focus === 'function') lastFocus.focus();
    lastFocus = null;
  }

  document.addEventListener('click', triggerFromEvent);
  document.addEventListener('keydown', e => {
    if (e.key !== 'Enter' && e.key !== ' ') return;
    if (!e.target.matches('[data-detail-domain][data-detail-id]')) return;
    e.preventDefault();
    triggerFromEvent(e);
  });

  window.CupolaDetails = { open, close };
})();
